src/mpi/coll/transports/gentran/gentran_utils.c

```diff
@@ -444,7 +447,7 @@ void MPII_Genutil_vtx_add_dependencies(MPII_Genutil_sched_t * sched, int vtx_id,
     /* check if there was any fence operation and add appropriate dependencies.
      * The application will never explicitly specify a dependency on it,
      * the transport has to make sure that the dependency on the fence operation is met */
-    if (sched->last_fence != -1 && sched->last_fence != vtx_id && n_in_vtcs == 0) {
+    if (sched->last_fence != -1 && sched->last_fence != vtx_id) {
```

Comment:
This was one of the important optimizations in gentran for performance to reduce dependence (and so reduce potential mallocs). Can you limit your changes in the algorithms themselves without this change?

Reply:
How do I add an operation that doesn't have any other dependencies other than the fence? 

What you are suggesting is there is algorithm that is using a fence, but it is also critical for later operations to not honor the fence. I wonder in that case why it is using a fence at all? It should be using a `MPIR_sched_sink`, right? Could you show me an example for me to better understand?

Reply:
If you add a vertex (operation) that doesn't have any dependencies, you just add without any in_vtcs. 
Fence is implicit, that is, when there is an active fence, all vertices added later is implicitly depending on it. You don't need to explicitly add dependence to the fence. 
So basically it is opposite to what you said, fence is always honored for all the vertices added after the fence.

Reply:
The central idea of this PR is to add dependency based on semantics, for example, Send depends on send buffer, Recv depends on receive buffer. And at the first round, the send buffer is automatically ready, so we encode it as `-1`. That is, we could add operation without any dependency (other than the fence) but `n_in_vtcs` isn't `0` because they are `-1` during first round.

How about we  calculate `n_real_in_vtcs` and use that?

Reply:
Oh, I remember why we need fence to be real fence. In this algorithm, the send buffer dependency could be real dependency on `in_place` `dt_copy` that is before the fence, and it is for correctness that the operation honor both the `dt_copy` dependency and fence. Admittedly, we could write the code in a way to check whether the dependency is after a fence, but that puts the fence into the algorithm parts. The idea for fence is to separate the parts of algorithm so one part doesn't concern with the other. I think we should look at the code that use fence in a leaky way, and understand why is that, because to me, that is definitely wrong use of fence, rather it should use sink

Reply:
I see your concern that you don't want to add extra fence dependency when it isn't necessary. Let me dig into it to see if we can achieve that without compromising semantics.

Reply:
I think calculating n_real_in_vtcs is probably fine for what you wanted to do. so replace "n_in_vtcs == 0"
by "n_real_in_vtcs == 0"?

Reply:
Do I understand it correctly: you have a vertex "a" which depends on vertex "b" before the fence, and the fence itself.
If this is the case, isn't it enough to just depend on the fence. Since fence has an implicit semantics that the fence depends on all vertices before the fence.

Reply:
> If this is the case, isn't it enough to just depend on the fence. Since fence has an implicit semantics that the fence depends on all vertices before the fence.

Yes, we can. But that makes code explicitly depends on the fence and across the fence, thus more difficult to maintain. I want the code ignorant of the fence.

Reply:
Just added a commit https://github.com/pmodels/mpich/pull/5620/commits/ba9c9012f6f54c1d17aef8b44a6b3f019ccbfe53 to avoid unnecessarily adding fence. Please review